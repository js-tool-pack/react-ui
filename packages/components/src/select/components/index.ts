export * from './Menu.option';
export * from './Selection.filter';
export * from './Selection.tags';
export * from './Selection';
export * from './Menu';
export * from './TabTrigger';
