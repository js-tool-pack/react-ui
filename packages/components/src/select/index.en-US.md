---
category: Components
title: Select
atomId: Select
demo:
  cols: 2
group:
  title: Data Input
---

Select 选择器。

## 代码演示

<!-- prettier-ignore -->
<code src="./demo/basic.tsx"></code>
<code src="./demo/multiple.tsx"></code>
<code src="./demo/size.tsx"></code>
<code src="./demo/clear.tsx"></code>
<code src="./demo/icon.tsx"></code>
<code src="./demo/filter.tsx"></code>
<code src="./demo/group.tsx"></code>
<code src="./demo/max-tag-count.tsx"></code>
<code src="./demo/custom-render.tsx"></code>
<code src="./demo/remote.tsx"></code>
<code src="./demo/status.tsx"></code>
<code src="./demo/slots.tsx"></code>
<code src="./demo/controller.tsx"></code>

## API

### SelectProps

| 属性              | 说明                                                                                        | 类型                                                               | 默认值    | 版本 |
| ----------------- | ------------------------------------------------------------------------------------------- | ------------------------------------------------------------------ | --------- | ---- |
| multiple          | 多选                                                                                        | boolean                                                            | --        | --   |
| value             | 指定当前选中的条目，多选时为一个数组                                                        | React.Key[] \| React.Key                                           | --        | --   |
| size              | 尺寸                                                                                        | 'small' \| 'medium' \| 'large'                                     | 'medium'  | --   |
| options           | 选项                                                                                        | (SelectOption \| SelectOptionGroup \| SelectDivider)[]             | --        | --   |
| disabled          | 禁用                                                                                        | boolean                                                            | --        | --   |
| clearable         | 显示清理按钮                                                                                | boolean                                                            | --        | --   |
| loading           | 加载中                                                                                      | boolean                                                            | --        | --   |
| placeholder       | 提示信息                                                                                    | string                                                             | 'select'  | --   |
| status            | 验证状态                                                                                    | 'warning' \| 'error'                                               | --        | --   |
| onFocus           | focus 回调                                                                                  | () => void                                                         | --        | --   |
| onBlur            | blur 回调                                                                                   | () => void                                                         | --        | --   |
| onClear           | 清理全部已选时的回调                                                                        | () => void                                                         | --        | --   |
| onSelect          | 选中一个选项时的回调                                                                        | (option: SelectOption) => void;                                    | --        | --   |
| onChange          | 选中的结果变化时的回调                                                                      | (value: React.Key[] \| React.Key, options: SelectOption[]) => void | --        | --   |
| maxTagCount       | 当 multiple 为 true 时，限定展示 tag 的数量，多的会收起来                                   | number                                                             | --        | --   |
| filterable        | 过滤输入开关                                                                                | boolean                                                            | --        | --   |
| filter            | 默认过滤选项 label 文字，传入 filter 后可自定义过滤方式；可脱离 filterable 使用             | (pattern: string, option: SelectOption) => boolean;                | --        | --   |
| ignoreComposition | 忽略输入法 Composition 状态                                                                 | boolean                                                            | true      | --   |
| remote            | 是否异步获取选项。如果开启了，那么 filter 不会对 options 生效。这个时候你在全权控制 options | boolean                                                            | --        | --   |
| onSearch          | 异步获取选项时的回调                                                                        | (pattern: string) => void                                          | --        | --   |
| header            | 窗体顶部插槽                                                                                | React.ReactNode                                                    | --        | --   |
| footer            | 窗体底部插槽                                                                                | React.ReactNode                                                    | --        | --   |
| empty             | 选项为空时显示的内容                                                                        | React.ReactNode                                                    | 'No Data' | --   |
| icon              | 右侧图标插槽                                                                                | React.ReactNode                                                    | --        | --   |
| controllerRef     | 控制器                                                                                      | React.Ref\<SelectControllerRef\>                                   | --        | --   |
| attrs             | html 标签属性                                                                               | Partial\<React.HTMLAttributes\<HTMLDivElement>>                    | --        | --   |

其他属性见 [Popover](./popover#api) 组件

### SelectControllerRef

| 属性  | 说明     | 类型       | 版本 |
| ----- | -------- | ---------- | ---- |
| focus | 获取焦点 | () => void | --   |
| blur  | 移除焦点 | () => void | --   |

### SelectOption

| 属性     | 说明                       | 类型                                                                               | 默认值 | 版本 |
| -------- | -------------------------- | ---------------------------------------------------------------------------------- | ------ | ---- |
| type     | 类型                       | 'option'                                                                           | --     | --   |
| value    | 对应唯一值                 | React.Key                                                                          | --     | --   |
| label    | tag 或者 option 显示的内容 | React.ReactNode \| ((selected: boolean, option: SelectOption) =\> React.ReactNode) | --     | --   |
| disabled | 禁用                       | boolean                                                                            | --     | --   |
| attrs    | html 标签属性              | Partial\<React.HTMLAttributes\<HTMLDivElement>>                                    | --     | --   |

### SelectOptionGroup

选项组属性。

| 属性     | 说明          | 类型                                                   | 默认值 | 版本 |
| -------- | ------------- | ------------------------------------------------------ | ------ | ---- |
| type     | 类型          | 'group'                                                | --     | --   |
| key      | 唯一 key      | React.Key                                              | --     | --   |
| children | 子选项        | (SelectOption \| SelectOptionGroup \| SelectDivider)[] | --     | --   |
| attrs    | html 标签属性 | Partial\<React.HTMLAttributes\<HTMLDivElement>>        | --     | --   |

### SelectDivider

分割线属性。

| 属性 | 说明     | 类型      | 默认值 | 版本 |
| ---- | -------- | --------- | ------ | ---- |
| type | 类型     | 'divider' | --     | --   |
| key  | 唯一 key | React.Key | --     | --   |

其他属性见 [Divider](./divider#api) 组件
