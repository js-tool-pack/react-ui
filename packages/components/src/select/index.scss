@use '../namespace' as Name;
@use '../popover/index' as Pop;

$r: Name.$select;
$option: #{$r}-option;
$tags: #{$r}-tags;
$menu: #{$r}-menu;
$filter: #{$r}-filter;
$tab-trigger: #{$r}-tab-trigger;
$empty: #{$r}-empty;
$selection: #{$r}-selection;
$balloon-content: #{Name.$word-balloon}__content;

.#{$r} {
  --t-select-border-color: var(--t-color-info-a-3);
  --t-select-box-shadow: 0 0 0 0 var(--t-color-primary-a-6);
  --t-select-bg-color: initial;
  --t-select-active-border-color: var(--t-color-primary-a-2);
  --t-select-active-box-shadow: 0 0 0 2px var(--t-color-primary-a-6);

  display: flex;
  align-items: center;
  padding: 2px var(--t-radius);
  min-width: 120px;
  min-height: var(--t-size);
  border: 1px solid var(--t-select-border-color);
  border-radius: var(--t-radius);
  background-color: var(--t-select-bg-color);
  box-shadow: var(--t-select-box-shadow);
  transition: all var(--t-transition-duration);
  box-sizing: border-box;
  cursor: pointer;
  &__dropdown {
    --t-select-transition-duration: calc(var(--t-transition-duration) / 2);
    @include Pop.pop-style(var(--t-select-transition-duration));

    .#{$balloon-content} {
      padding: 4px;
    }
    .#{$option} {
      --t-option-border-radius: calc(var(--t-radius) / 2);
    }
  }
  &__dropdown-enter-from,
  &__dropdown-leave-to {
    transform: scale(0.9);
  }

  // :
  &:hover {
    &:not(.#{$r}--disabled) {
      border-color: var(--t-select-active-border-color);
    }
    &.#{$r}--selected.#{$r}--clearable {
      .#{$selection}__clear {
        display: block;
      }
      .#{$selection}__icon {
        opacity: 0;
      }
    }
  }

  // status
  &--active,
  &--focus {
    border-color: var(--t-select-active-border-color);
    box-shadow: var(--t-select-active-box-shadow);
  }
  &--warning {
    --t-select-border-color: var(--t-color-warning-a-2);
    --t-select-active-border-color: var(--t-color-warning-a-2);
    --t-select-active-box-shadow: 0 0 0 2px var(--t-color-warning-a-6);
  }
  &--error {
    --t-select-border-color: var(--t-color-danger-a-2);
    --t-select-active-border-color: var(--t-color-danger-a-2);
    --t-select-active-box-shadow: 0 0 0 2px var(--t-color-danger-a-6);
  }
  &--disabled {
    --t-select-bg-color: var(--t-color-info-a-7);
    --t-select-border-color: var(--t-color-info-a-3);

    cursor: not-allowed;
  }
}

.#{$tags} {
  &__pop-tags {
    display: inline-flex;
    align-items: center;
    overflow: hidden;
    flex: 1;
    flex-wrap: wrap;
    gap: 4px;
    max-width: 30vw;
  }
}
.#{$menu} {
  &__body {
    position: relative;
    overflow: auto;
    max-height: 50vh;
    ::-webkit-scrollbar {
      margin-right: 4px;
      width: 6px;
      height: 6px;
      background-color: transparent;
    }
    ::-webkit-scrollbar-corner {
      display: none;
    }
    ::-webkit-scrollbar-thumb {
      border-radius: 4px;
      background-color: rgb(220 220 242 / 18%);
    }
  }
  &__group-title {
    --t-option-color: var(--t-option-diabled-color);
  }
  &__options,
  &__group-body {
    margin: 0;
    padding: 0;
    list-style: none;
  }
  &__divider {
    margin: 4px 0;
  }
  &__group-body {
    margin-left: var(--t-radius);
  }
}
.#{$empty} {
  &__content {
    text-align: center;
    line-height: 60px;
    color: var(--t-color-info);
  }
}
.#{$filter} {
  display: inline-block;
  overflow: hidden;
  &__input {
    overflow: hidden;
    padding: 0;
    min-width: 1px;
    max-width: 100%;
    border: 0;
    color: var(--t-text-color);
    background-color: #00000000;
    outline: 0;
  }
  &__mirror {
    position: absolute;
    top: 0;
    left: 0;
    white-space: pre;
    opacity: 0;
    visibility: hidden;
    user-select: none;
  }
}
.#{$selection} {
  position: relative;
  width: 100%;
  &,
  &__selected,
  &__suffix {
    display: flex;
    align-items: center;
  }
  &__selected {
    display: inline-flex;
    align-items: center;
    overflow: hidden;
    flex: 1;
    flex-wrap: wrap;
    gap: 4px;
  }
  &__placeholder {
    position: absolute;
    color: var(--t-color-info-a-3);
  }
  &__suffix {
    position: relative;
    padding: 0 4px;
  }
  &__icon {
    font-size: 0.8em;
  }
  &__clear {
    position: absolute;
    cursor: pointer;
    display: none;
    color: var(--t-color-info-a-4);
    &:hover {
      color: var(--t-color-info);
    }
  }
}
.#{$tab-trigger} {
  padding: 0;
  width: 0;
  border: 0;
  opacity: 0;
  outline: 0;
}
.#{$option} {
  &--selected {
    --t-option-bg-color: var(--t-color-primary-a-7);

    color: var(--t-color-primary-a-2);
  }
  &--picked {
    --t-option-bg-color: var(--t-option-active-bg-color);
  }
  &__selected-icon {
    font-size: 0.8em;
  }
}
