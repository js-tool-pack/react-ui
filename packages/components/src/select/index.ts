export type {
  SelectControllerRef,
  SelectStaticProps,
  SelectOptionsItem,
  SelectDivider,
  SelectOption,
  SelectLocale,
  SelectProps,
} from './select.types';
export * from './Select';
