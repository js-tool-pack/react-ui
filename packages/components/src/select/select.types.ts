import { DividerProps } from '~/divider';
import React from 'react';
import { OptionProps } from '~/option';
import { PopoverProps } from '~/popover';
import { Size } from '@pkg/shared';

export interface SelectOptionProps extends OptionProps {
  selected?: boolean;
  picked?: boolean;
}
export interface SelectDivider extends DividerProps {
  type: 'divider';
  key: React.Key;
}
export interface SelectOption extends Omit<OptionProps, 'size' | 'children'> {
  type?: 'option';
  value: React.Key;
  label:
    | React.ReactNode
    | ((selected: boolean, option: SelectOption) => React.ReactNode);
  // [key: string]: unknown;
}
export interface SelectOptionGroup
  extends Omit<OptionProps, 'size' | 'children' | 'disabled' | 'readonly'> {
  type: 'group';
  key: React.Key;
  label: React.ReactNode;
  children?: SelectOptionsItem[];
}
export type SelectOptionsItem =
  | SelectOption
  | SelectOptionGroup
  | SelectDivider;

export interface SelectControllerRef {
  focus(): void;
  blur(): void;
}

export interface SelectStaticProps
  extends Omit<PopoverProps, 'content' | 'name' | 'children'> {
  popoverAttrs?: Partial<React.HTMLAttributes<HTMLDivElement>>;

  size?: Size;
  options: SelectOptionsItem[];
  disabled?: boolean;
  clearable?: boolean;
  loading?: boolean;
  placeholder?: string;
  status?: 'warning' | 'error';
  onFocus?: () => void;
  onBlur?: () => void;
  onClear?: () => void;
  onSelect?: (option: SelectOption) => void;
  maxTagCount?: number;
  ignoreComposition?: boolean;

  // 过滤
  filterable?: boolean;
  filter?: (pattern: string, option: SelectOption) => boolean;

  // 远程
  remote?: boolean;
  onSearch?: (pattern: string) => void;

  // slots
  header?: React.ReactNode;
  footer?: React.ReactNode;
  empty?: React.ReactNode;
  icon?: React.ReactNode;

  // 控制器
  controllerRef?: React.Ref<SelectControllerRef>;

  // dynamic
  value?: React.Key[] | React.Key;
  multiple?: boolean;
  onChange?: (value: any, options: SelectOption[]) => void;
}
export interface SelectProps<
  Multiple extends boolean = false,
  ValueType extends React.Key = React.Key,
> extends SelectStaticProps {
  value?: Multiple extends true ? ValueType[] : ValueType;
  multiple?: Multiple;
  onChange?: (
    value: Multiple extends true ? ValueType[] : ValueType,
    options: SelectOption[],
  ) => void;
}

export type SelectFC = <
  Multiple extends boolean = false,
  ValueType extends React.Key = React.Key,
>(
  props: SelectProps<Multiple, ValueType>,
) => React.ReactElement;
