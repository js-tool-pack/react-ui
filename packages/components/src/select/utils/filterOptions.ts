import type { SelectOption, SelectOptionsItem } from '../select.types';
import { isString } from '@tool-pack/basic';
import { isDivider } from './isDivider';
import { isGroup } from './isGroup';

export function filterOptions(
  options: SelectOptionsItem[],
  pattern: string,
  filter = filterLabel,
): SelectOptionsItem[] {
  const result: SelectOptionsItem[] = [];
  options.forEach((opt) => {
    if (isDivider(opt)) return;
    if (isGroup(opt)) {
      if (!opt.children?.length) return;
      const children = filterOptions(opt.children, pattern, filter);
      if (!children.length) return;
      result.push({ ...opt, children });
      return;
    }
    if (filter(pattern, opt)) result.push(opt);
  });
  return result;
}

function filterLabel(pattern: string, option: SelectOption): boolean {
  return isString(option.label) && option.label.includes(pattern);
}
