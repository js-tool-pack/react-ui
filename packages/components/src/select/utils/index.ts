export * from './pickOptionsByValue';
export * from './flattenOptions';
export * from './isValueChanged';
export * from './filterOptions';
export * from './isDivider';
export * from './intoView';
export * from './isGroup';
