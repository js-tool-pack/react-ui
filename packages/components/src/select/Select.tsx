import React, { useEffect, useImperativeHandle, useRef, useState } from 'react';
import type {
  SelectOptionsItem,
  SelectStaticProps,
  SelectOption,
  SelectProps,
  SelectFC,
} from './select.types';
import {
  useEventListenerOnMounted,
  useStateWithTrailClear,
  getSizeClassName,
  useForwardRef,
  getClasses,
} from '@pkg/shared';
import type { RequiredPart } from '@tool-pack/types';
import { Popover } from '~/popover';
import { Menu, Selection, TabTrigger } from '~/select/components';
import { getClassNames } from '@tool-pack/basic';
import {
  pickOptionsByValue,
  isValueChanged,
  filterOptions,
} from '~/select/utils';
import { useEsc } from '~/dialog/dialog.hooks';
import { filter as rxFilter, fromEvent } from 'rxjs';
import { transitionCBAdapter } from '~/transition';

const cls = getClasses(
  'select',
  [],
  ['active', 'selected', 'disabled', 'clearable', 'focus'],
);
const defaultProps = {
  ignoreComposition: true,
  placeholder: 'select',
  trigger: 'click',
  widthByTrigger: true,
  showArrow: false,
  placement: 'bottom',
  offset: 2,
  multiple: false,
  size: 'medium',
} satisfies Partial<SelectStaticProps>;

const _Select: React.FC<SelectStaticProps> = React.forwardRef<
  HTMLDivElement,
  SelectStaticProps
>((props, ref) => {
  const {
    ignoreComposition,
    onVisibleChange,
    controllerRef,
    popoverAttrs,
    placeholder,
    maxTagCount,
    filterable,
    clearable,
    showArrow,
    disabled,
    onChange,
    onSelect,
    onSearch,
    multiple,
    onClear,
    loading,
    options,
    visible,
    onFocus,
    onBlur,
    header,
    filter,
    remote,
    empty,
    footer,
    status,
    value,
    icon,
    size,
    attrs = {},
    ...rest
  } = props as RequiredPart<SelectProps, keyof typeof defaultProps>;

  const rootRef = useForwardRef(ref);
  const isTrailFocusRef = useRef(false);
  const tabTriggerRef = useRef<HTMLInputElement>(null);
  const [opened, setOpened] = useState(false);
  const [focus, setFocus] = useState<boolean>();
  const [show, setShow] = useStateWithTrailClear(visible);
  const [selected, setSelected] = useState<SelectOption[]>([]);
  const [pattern, setPattern] = useState('');

  useImperativeHandle(controllerRef, () => {
    return {
      focus() {
        tabTriggerRef.current?.focus();
      },
      blur() {
        tabTriggerRef.current?.blur();
      },
    };
  });

  // 页面 blur
  useEventListenerOnMounted(
    window,
    'blur',
    close,
    undefined,
    // active,
    false,
  );

  // 单选关闭窗体
  useEffect(() => {
    if (!multiple && opened) closeWithFocus();
    // 多选时选过后清理过滤值
    if (multiple && (filterable || remote)) setPattern('');
  }, [selected]);

  // 回显
  useEffect(() => {
    if (isValueChanged(value, selected)) {
      setSelected(pickOptionsByValue(options, value));
    }
  }, [value]);

  useEffect(() => {
    if (!opened) return;

    // 监听 tab 键
    const sub$ = fromEvent<KeyboardEvent>(window, 'keydown')
      .pipe(rxFilter((e) => e.code === 'Tab'))
      .subscribe(close);

    return () => sub$.unsubscribe();
  }, [opened]);

  useEffect(() => {
    if (focus === undefined) return;
    if (focus) onFocus?.();
    else onBlur?.();
  }, [focus]);

  useEsc(opened, true, closeWithFocus);

  // 过滤选项
  const filteredOptions = getFilteredOptions();

  const popoverOn = transitionCBAdapter({
    onBeforeEnter: () => {
      setOpened(true);
      onVisibleChange?.(true);
      setFocus(true);
    },
    onBeforeLeave: () => {
      setOpened(false);
      onVisibleChange?.(false);
    },
    onAfterLeave() {
      // 关闭后清理过滤值
      if (!multiple) setPattern('');
      if (isTrailFocusRef.current) {
        tabTriggerRef.current?.focus();
        isTrailFocusRef.current = false;
      } else {
        setFocus(false);
      }
    },
  });

  return (
    <Popover
      {...rest}
      on={popoverOn}
      name="select__dropdown"
      showArrow={showArrow}
      attrs={popoverAttrs}
      disabled={disabled}
      visible={show}
      content={
        <Menu
          onSelectedChange={onSelectedChange}
          options={filteredOptions}
          selected={selected}
          multiple={multiple}
          onSelect={onSelect}
          header={header}
          footer={footer}
          active={opened}
          empty={empty}
        />
      }>
      {/* eslint-disable-next-line jsx-a11y/click-events-have-key-events,jsx-a11y/no-static-element-interactions */}
      <div
        {...attrs}
        className={getClassNames(
          cls.root,
          attrs.className,
          getSizeClassName(size),
          {
            [`${cls.root}--${status}`]: status,
            [cls['--'].selected]: selected.length,
            [cls['--'].clearable]: clearable,
            [cls['--'].disabled]: disabled,
            [cls['--'].active]: opened,
            [cls['--'].focus]: focus,
          },
        )}
        onClick={onRootClick}
        ref={rootRef}>
        <TabTrigger
          disabled={disabled}
          ref={tabTriggerRef}
          onFocus={_onFocus}
          onBlur={_onBlur}
          opened={opened}
          onOpen={open}
        />
        <Selection
          ignoreComposition={ignoreComposition}
          onPatternChange={onPatternChange}
          onSelectedChange={onSelectedChange}
          maxTagCount={maxTagCount}
          placeholder={placeholder}
          filterable={filterable}
          clearable={clearable}
          disabled={disabled}
          multiple={multiple}
          selected={selected}
          pattern={pattern}
          loading={loading}
          options={options}
          onClear={onClear}
          opened={opened}
          remote={remote}
          icon={icon}
          size={size}
        />
      </div>
    </Popover>
  );

  function _onFocus(): void {
    setFocus(true);
  }
  function _onBlur(): void {
    if (opened || isTrailFocusRef.current) return;
    setFocus(false);
  }
  function onSelectedChange(selected: SelectOption[]): void {
    setSelected(selected);
    if (!onChange) return;
    const selectedValues = selected.map((s) => s.value);
    const value = (multiple ? selectedValues : selectedValues[0]) as Parameters<
      typeof onChange
    >[0];
    onChange(value, selected);
  }
  function getFilteredOptions(): SelectOptionsItem[] {
    // remote 由外部决定选项是什么，对应的回调是 onSearch，避免双重过滤
    if (remote) return options;
    if (!pattern && !filter) return options;
    return filterOptions(options, pattern, filter);
  }
  function onPatternChange(value: string): void {
    setPattern(value);
    if (remote) onSearch?.(value);
  }
  function closeWithFocus(): void {
    isTrailFocusRef.current = true;
    close();
  }
  function close(): void {
    setOpened(false);
    setShow(false);
  }
  function open(): void {
    setOpened(true);
    setShow(true);
  }
  function onRootClick(e: React.MouseEvent<HTMLDivElement>): void {
    attrs.onClick?.(e);
    if (disabled || !opened) return;
    if (filterable || remote) {
      e.stopPropagation();
      e.preventDefault();
      return;
    }
    isTrailFocusRef.current = true;
  }
});

_Select.defaultProps = defaultProps;
_Select.displayName = 'Select';

export const Select = _Select as SelectFC;
