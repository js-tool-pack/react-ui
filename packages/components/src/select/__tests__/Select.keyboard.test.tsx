import { fireEvent, render, act } from '@testing-library/react';
import { SelectOption, Select } from '~/select';
import { getBalloon, $$, $ } from './utils';

describe('Select.keyboard', () => {
  const options = [
    {
      label: 'foo',
      value: 1,
    },
    {
      label: 'bar',
      value: 2,
    },
    {
      label: 'foo bar',
      value: 3,
    },
  ] satisfies SelectOption[];

  const cls = {
    selected: 't-select-option--selected',
    invisible: 't-transition--invisible',
    picked: 't-select-option--picked',
    active: 't-input-skin--active',
  };

  it('should close Select when esc keydown', () => {
    jest.useFakeTimers();
    const { container } = render(<Select options={options} visible />);
    expect(container.firstChild).toHaveClass(cls.active);
    expect(getBalloon()).not.toHaveClass(cls.invisible);

    fireEvent.keyDown(window, { key: 'Escape' });
    act(() => jest.advanceTimersByTime(500));

    expect(container.firstChild).toHaveClass(cls.active);
    expect(getBalloon()).toHaveClass(cls.invisible);
  });

  it('should highlight the first option', () => {
    render(<Select options={options} visible />);

    const $options = $$('.t-select-option');
    expect($options.length).toBe(3);

    expect($options[0]).toHaveClass(cls.picked);
    expect($options[1]).not.toHaveClass(cls.picked);
    expect($options[2]).not.toHaveClass(cls.picked);
  });

  it('should highlight the first selected option', () => {
    const value = options[1]!.value;
    render(<Select options={options} value={value} visible />);

    const $options = $$('.t-select-option');
    expect($options.length).toBe(3);

    expect($options[0]).not.toHaveClass(cls.picked);
    expect($options[1]).toHaveClass(cls.picked);
    expect($options[2]).not.toHaveClass(cls.picked);
  });

  it('should highlight the first option when tab-trigger trigger', () => {
    jest.useFakeTimers();
    render(<Select options={options} />);

    const tabTrigger = $('.t-input-popover-tab-trigger')!;
    expect(getBalloon()).toBeNull();

    act(() => tabTrigger.focus());
    expect(tabTrigger).toHaveFocus();

    fireEvent.keyDown(tabTrigger, { code: 'Enter' });
    expect(getBalloon()).not.toBeNull();

    act(() => jest.advanceTimersByTime(500));

    const $options = $$('.t-select-option');
    expect($options.length).toBe(3);

    expect($options[0]).toHaveClass(cls.picked);
    expect($options[1]).not.toHaveClass(cls.picked);
    expect($options[2]).not.toHaveClass(cls.picked);
  });

  it("should allow controlling the selected option using the keyboard's up and down arrow keys and selecting by pressing the Enter key", () => {
    jest.useFakeTimers();
    const { container } = render(<Select options={options} visible />);

    const $options = $$('.t-select-option');

    expect(getBalloon()).not.toHaveClass(cls.invisible);
    expect($options.length).toBe(3);

    expect($options[0]).not.toHaveClass(cls.selected);
    expect($options[1]).not.toHaveClass(cls.selected);
    expect($options[2]).not.toHaveClass(cls.selected);

    expect($options[0]).toHaveClass(cls.picked);
    expect($options[1]).not.toHaveClass(cls.picked);
    expect($options[2]).not.toHaveClass(cls.picked);

    fireEvent.keyDown(window, { code: 'ArrowDown' });
    expect($options[0]).not.toHaveClass(cls.picked);
    expect($options[1]).toHaveClass(cls.picked);
    expect($options[2]).not.toHaveClass(cls.picked);

    fireEvent.keyDown(window, { code: 'ArrowDown' });
    expect($options[0]).not.toHaveClass(cls.picked);
    expect($options[1]).not.toHaveClass(cls.picked);
    expect($options[2]).toHaveClass(cls.picked);

    fireEvent.keyDown(window, { code: 'ArrowDown' });
    expect($options[0]).toHaveClass(cls.picked);
    expect($options[1]).not.toHaveClass(cls.picked);
    expect($options[2]).not.toHaveClass(cls.picked);

    fireEvent.keyDown(window, { code: 'ArrowUp' });
    expect($options[0]).not.toHaveClass(cls.picked);
    expect($options[1]).not.toHaveClass(cls.picked);
    expect($options[2]).toHaveClass(cls.picked);

    fireEvent.keyDown(window, { code: 'ArrowUp' });
    expect($options[0]).not.toHaveClass(cls.picked);
    expect($options[1]).toHaveClass(cls.picked);
    expect($options[2]).not.toHaveClass(cls.picked);

    fireEvent.keyDown(window, { code: 'Enter' });
    expect($options[0]).not.toHaveClass(cls.picked);
    expect($options[1]).toHaveClass(cls.picked + ' ' + cls.selected);
    expect($options[2]).not.toHaveClass(cls.picked);

    expect(container.querySelector('.t-select__selected')).toMatchSnapshot();

    expect(getBalloon()).not.toHaveClass(cls.invisible);
    act(() => jest.advanceTimersByTime(500));
    expect(getBalloon()).toHaveClass(cls.invisible);
  });
});
