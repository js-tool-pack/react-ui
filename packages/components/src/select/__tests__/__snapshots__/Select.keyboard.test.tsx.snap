// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select.keyboard should allow controlling the selected option using the keyboard's up and down arrow keys and selecting by pressing the Enter key 1`] = `null`;
